spherical_splines/SHTOOLS/man/man1/makegriddh.1
.\" Automatically generated by Pod::Man 2.16 (Pod::Simple 3.05)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sh \" Subsection heading
.br
.if t .Sp
.ne 5
.PP
\fB\\$1\fR
.PP
..
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.Sh), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.ie \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    nr % 0
.    rr F
.\}
.el \{\
.    de IX
..
.\}
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "MAKEGRIDDH 1"
.TH MAKEGRIDDH 1 "2009-08-21" "SHTOOLS 2.6" "SHTOOLS 2.6"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "MakeGridDH"
.IX Header "MakeGridDH"
.IP "MakeGridDH \-" 14
.IX Item "MakeGridDH -"
Create a 2D map from a set of spherical harmonic coefficients that conforms with Driscoll and Healy's (1994) sampling theorem.
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
.IP "\s-1SUBROUTINE\s0 MakeGridDH (" 24
.IX Item "SUBROUTINE MakeGridDH ("
\&\s-1GRIDDH\s0, N, \s-1CILM\s0, \s-1LMAX\s0, \s-1NORM\s0, \s-1SAMPLING\s0, \s-1CSPHASE\s0, \s-1LMAX_CALC\s0 )
.RS 4
.IP "REAL*8" 19
.IX Item "REAL*8"
\&\s-1GRIDDH\s0(2*LMAX+2, 2*LMAX+2), \s-1CILM\s0(2,\ \s-1LMAX+1\s0,\ \s-1LMAX+1\s0)
.IP "\s-1INTEGER\s0" 19
.IX Item "INTEGER"
\&\s-1LMAX\s0, N
.IP "\s-1INTEGER\s0, \s-1OPTIONAL\s0" 19
.IX Item "INTEGER, OPTIONAL"
\&\s-1NORM\s0, \s-1SAMPLING\s0, \s-1CSPHASE\s0, \s-1LMAX_CALC\s0
.RE
.RS 4
.RE
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
\&\fIMakeGridDH\fR will create a 2\-dimensional map equally sampled (N by N) or equally spaced (N by 2N) in latitude and longitude from a set of input spherical harmonic coefficients, where N is 2*LMAX+2. This grid conforms with the sampling theorem of Driscoll and Healy (1994) and this routine is the inverse of \fISHExpandDH\fR. The function is evaluated at each longitudinal band by inverse Fourier transforming the \s-1SIN\s0 and \s-1COS\s0 terms for each degree l, and then summing over all degrees. When evaluating the function, the maximum spherical harmonic degree that is considered is the minimum of \s-1LMAX\s0, the size of \s-1CILM\-1\s0, or \s-1LMAX_CALC\s0 (if specified).
.PP
The default is to use an input grid that is equally sampled (N by N), but this can be changed to use an equally spaced grid (N by 2N) by the optional argument \s-1SAMPLING\s0. The first latitudinal band of the grid corresponds to 90 N, the latitudinal band for 90 S is not calculated, and the latitudinal sampling interval is 180/N degrees. The first longitudinal band is 0 E, the longitudinal band for 360 E is not calculated, and the longitudinal sampling interval is 360/N for equally sampled and 180/N for equally spaced grids, respectively.
.PP
The employed spherical harmonic normalization and Condon-Shortley phase convention can be set by the optional arguments \s-1NORM\s0 and \s-1CSPHASE\s0; if not set, the default is to use geodesy 4\-pi normalized harmonics that exclude the Condon-Shortley phase of (\-1)^m.
.SH "ARGUMENTS"
.IX Header "ARGUMENTS"
.IP "\s-1GRIDDH\s0" 12
.IX Item "GRIDDH"
(output) REAL*8, \s-1DIMENSION\s0(2*LMAX+2, 2*LMAX+2) \fIor\fR
(output) REAL*8, \s-1DIMENSION\s0(2*LMAX+2, 4*LMAX+4)
.Sp
A 2D equally sampled (N by N; default), or equally spaced (N by 2N) map of the input spherical harmonic coefficients \s-1CILM\s0 that conforms to the sampling theorem of Driscoll and Healy (1994). The first latitudinal band corresponds to 90 N, the latitudinal band for 90 S is not included, and the latitudinal sampling interval is 180/N degrees. The first longitudinal band is 0 E, the longitudinal band for 360 E is not included, and the longitudinal sampling interval is 360/N for an equally sampled and 180/N for an equally spaced grid, respectively.
.IP "N" 12
.IX Item "N"
(output) \s-1INTEGER\s0
.Sp
The number of samples in latitude and longitude of \s-1GRIDDH\s0. This is equal to 2*LMAX + 2, which will always be even.
.IP "\s-1CILM\s0" 12
.IX Item "CILM"
(input) REAL*8, \s-1DIMENSION\s0 (2, \s-1LMAX+1\s0, \s-1LMAX+1\s0)
.Sp
The real spherical harmonic coefficients of the function. The coefficients C1lm and C2lm refer to the \*(L"cosine\*(R" (Clm) and \*(L"sine\*(R" (Slm) coefficients, respectively, with Clm = \s-1CILM\s0(1, l+1, m+1) and Slm = \s-1CILM\s0(2, l+1, m+1). Alternatively, C1lm and C2lm correspond to the positive and negative order coefficients, respectively.
.IP "\s-1LMAX\s0" 12
.IX Item "LMAX"
(input) \s-1INTEGER\s0
.Sp
The maximum spherical harmonic degree of the function. This determines the number of samples N.
.IP "\s-1NORM\s0" 12
.IX Item "NORM"
(input) \s-1INTEGER\s0, \s-1OPTIONAL\s0
.Sp
1 (default) = Geodesy 4\-pi normalized harmonics; 2 = Schmidt semi-normalized harmonics; 3 = unnormalized harmonics; 4 = orthonormal harmonics.
.IP "\s-1SAMPLING\s0" 12
.IX Item "SAMPLING"
(input) \s-1INTEGER\s0, \s-1OPTIONAL\s0
.Sp
If 1 (default) the input grid is equally sampled (N by N). If 2, the grid is equally spaced (N by 2*N).
.IP "\s-1CSPHASE\s0" 12
.IX Item "CSPHASE"

spherical_splines/SHTOOLS/src/MakeGridDH.f95
subroutine MakeGridDH(griddh, n, cilm, lmax, norm, sampling, csphase, lmax_calc)
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
!
!	Given the Spherical Harmonic coefficients CILM, this subroutine
!	will evalate the function on a grid with an equal number of samples N in 
!	both latitude and longitude (or N by 2N by specifying the optional parameter
!	SAMPLING = 2). This is the inverse of the routine SHExpandDH, both of which
!	are done quickly using FFTs for each degree of each latitude band. 
!	The number of samples is determined by the spherical harmonic bandwidth LMAX.
!	Nevertheless, the coefficients can be evaluated up to smaller spherical harmonic 
!	degree by specifying the optional parameter LMAX_CALC. Note that N is always 
!	EVEN for this routine. 
!	
!	The Legendre functions are computed on the fly using the scaling methodolgy 
!	presented in Holmes and Featherston (2002). When NORM = 1, 2 or 4, these are 
!	accurate to about degree 2800. When NORM = 3, the routine is only stable to about 
!	degree 15!
!
!	The output grid contains N samples in latitude from 90 to -90+interval, and in 
!	longitude from 0 to 360-2*interval (or N x 2N, see below), where interval is the 
!	sampling interval, and n=2*(LMAX+1). Note that the datum at 90 degees latitude 
!	is ultimately downweighted to zero, so this point does not contribute to the 
!	spherical harmonic coefficients.
!
!	Calling Parameters
!		IN
!			cilm		Input spherical harmonic coefficients with 
!					dimension (2, lmax+1, lmax+1).
!			lmax		Maximum spherical harmonic degree used in the expansion.
!					This determines the spacing of the output grid.
!		OUT
!			griddh		Gridded data of the spherical harmonic
!					coefficients CILM with dimensions (2*LMAX+2 , 2*LMAX+2). 
!			n		Number of samples in the grid, always even, which is 2*(LMAX+2).
!		OPTIONAL (IN)
!			norm:		Normalization to be used when calculating Legendre functions
!						(1) "geodesy" (default)
!						(2) Schmidt
!						(3) unnormalized
!						(4) orthonormalized
!			sampling	(1) Grid is N latitudes by N longitudes (default).
!					(2) Grid is N by 2N. The higher frequencies resulting
!					from this oversampling in longitude are discarded, and hence not
!					aliased into lower frequencies.
!			csphase		1: Do not include the phase factor of (-1)^m
!					-1: Apply the phase factor of (-1)^m.
!			lmax_calc	The maximum spherical harmonic degree to evaluate
!					the coefficients up to.
!
!	Notes:
!		1.	If lmax is greater than the the maximum spherical harmonic
!			degree of the input file, then this file will be ZERO PADDED!
!			(i.e., those degrees after lmax are assumed to be zero).
!		2. 	Latitude is geocentric latitude.
!
!	Dependencies:	FFTW3, CSPHASE_DEFAULT
!
!	Written by Mark Wieczorek 2004.
!
!	September 3, 2005. 	Modifed so that the array plx is now optional.
!	September 26, 2005. 	Added optional argument NORM.
!	May 30, 2006. 		Modified routine to take into acount the symmetry of the
!				Legendre functions about the equator, i.e., (-1)^(l+m).
!	October 16, 2006. 	The Legendre functions are now computed within this program
!				during the summations over l and m. This leads to an increase in speed
!				of about a factor of 2. Added optional argument SAMPLING.
!	July 22, 2007.		Added optional parameter LMAX_CALC for evaluating coefficients up
!				to a maximum degree smaller than LMAX. 
!	November 21, 2011	Fixed problem where saved variables used in Plm recursion were not recalculated
!				if NORM changed from one call to the next (with the same value of N).
!
!	Copyright (c) 2006-2011, Mark A. Wieczorek
!	All rights reserved.
!
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
	use FFTW3
	use SHTOOLS, only: CSPHASE_DEFAULT
	
	implicit none
	
	real*8, intent(in) :: 	cilm(:,:,:)
	real*8, intent(out) ::	griddh(:,:)
	integer, intent(in) :: 	lmax
	integer, intent(out) ::	n
	integer, intent(in), optional :: norm, sampling, csphase, lmax_calc
	integer :: 		l, m, i, l1, m1, lmax_comp, i_eq, i_s, phase, astat(4), lnorm, &
				k, kstart, nlong
	real*8 :: 		grid(4*lmax+4), pi, theta, coef0, scalef, rescalem, u, p, pmm, &
				pm1, pm2, z, coef0s, tempr
	complex*16 ::		coef(2*lmax+3), coefs(2*lmax+3), tempc
	integer*8 ::		plan
	real*8, save, allocatable ::	f1(:), f2(:), sqr(:), symsign(:)
	integer, save ::	lmax_old=0, norm_old = 0
	
	n = 2*lmax+2
	
	if (present(sampling)) then
		if (sampling /= 1 .and. sampling /=2) then
			print*, "Error --- MakeGridDH"
			print*, "Optional parameter SAMPLING must be 1 (N by N) or 2 (N by 2N)."
			print*, "Input value is ", sampling
			stop
		endif
	endif
	
	if (size(cilm(:,1,1)) < 2) then
		print*, "Error --- MakeGridDH"
		print*, "CILM must be dimensioned as (2, *, *)."
		print*, "Input dimension is ", size(cilm(:,1,1)), size(cilm(1,:,1)), size(cilm(1,1,:))
		stop
	endif 
	
	if (present(sampling)) then
		if (sampling == 1) then
			if (size(griddh(:,1)) < n .or. size(griddh(1,:)) < n) then
				print*, "Error --- MakeGridDH"
				print*, "GRIDDH must be dimensioned as (N, N) where N is ", n
				print*, "Input dimension is ", size(griddh(:,1)), size(griddh(1,:))
				stop
			endif
		elseif (sampling ==2) then
			if (size(griddh(:,1)) < n .or. size(griddh(1,:)) < 2*n) then
				print*, "Error --- MakeGriddDH"
				print*, "GRIDDH must be dimensioned as (N, 2*N) where N is ", n
				print*, "Input dimension is ", size(griddh(:,1)), size(griddh(1,:))
				stop
			endif
		endif
	else
		
		if (size(griddh(:,1)) < n .or. size(griddh(1,:)) < n) then
			print*, "Error --- MakeGridDH"
			print*, "GRIDDH must be dimensioned as (N, N) where N is ", n
			print*, "Input dimension is ", size(griddh(:,1)), size(griddh(1,:))
			stop
		endif

	endif
		
	if (present(norm)) then
		if (norm > 4 .or. norm < 1) then
			print*, "Error --- MakeGridDH"
			print*, "Parameter NORM must be 1 (geodesy), 2 (Schmidt), 3 (unnormalized), or 4 (orthonormalized)."
			print*, "Input value is ", norm
			stop
		endif
		
		lnorm = norm
		
	else
		lnorm = 1
	endif
	
	if (present(csphase)) then
     		if (csphase /= -1 .and. csphase /= 1) then
     			print*, "MakeGridDH --- Error"
     			print*, "CSPHASE must be 1 (exclude) or -1 (include)"
     			print*, "Input valuse is ", csphase
     			stop
     		else
     			phase = csphase
     		endif
     	else
     		phase = CSPHASE_DEFAULT
     	endif

	pi = acos(-1.0d0)
	
	scalef = 1.0d-280
	
	if (present(lmax_calc)) then
		lmax_comp = min(lmax, size(cilm(1,1,:))-1, size(cilm(1,:,1))-1, lmax_calc)
	else
		lmax_comp = min(lmax, size(cilm(1,1,:))-1, size(cilm(1,:,1))-1)
	endif
	
	if (present(sampling)) then
		if (sampling == 1) then
			nlong = n
		else
			nlong = 2*n
		endif
	else
		nlong = n
	endif
	
	!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
	!
	!	Calculate recursion constants used in computing the Legendre polynomials
	!
	!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
	
	if (lmax_comp /= lmax_old .or. lnorm /= norm_old) then
		
		if (allocated(sqr)) deallocate(sqr)
		if (allocated(f1)) deallocate(f1)
		if (allocated(f2)) deallocate(f2)
		if (allocated(symsign)) deallocate(symsign)
		
		allocate(sqr(2*lmax_comp+1), stat=astat(1))
		allocate(f1((lmax_comp+1)*(lmax_comp+2)/2), stat=astat(2))
		allocate(f2((lmax_comp+1)*(lmax_comp+2)/2), stat=astat(3))
		allocate(symsign((lmax_comp+1)*(lmax_comp+2)/2), stat=astat(4))
		
		if (astat(1) /= 0 .or. astat(2) /= 0 .or. astat(3) /= 0 .or. astat(4) /= 0) then
			print*, "MakeGridDH --- Error"
			print*, "Problem allocating arrays SQR, F1, F2, or SYMSIGN", astat(1), astat(2), astat(3), astat(4)
			stop
		endif
		
		!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
		!
		! 	Calculate signs used for symmetry of Legendre functions about equator
		!
		!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
		
		k = 0
		do l = 0, lmax_comp, 1
			do m = 0, l, 1
				k = k + 1
				if (mod(l-m,2) == 0) then
					symsign(k) = 1.0d0
				else
					symsign(k) = -1.0d0
				endif
			enddo
		enddo
			
		!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
		!
		!	Precompute square roots of integers that are used several times.
		!
		!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
	
		do l=1, 2*lmax_comp+1
			sqr(l) = sqrt(dble(l))
		enddo

		!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
		!
		! 	Precompute multiplicative factors used in recursion relationships
		! 		P(l,m) = x*f1(l,m)*P(l-1,m) - P(l-2,m)*f2(l,m)
		!		k = l*(l+1)/2 + m + 1
		!	Note that prefactors are not used for the case when m=l as a different 
		!	recursion is used. Furthermore, for m=l-1, Plmbar(l-2,m) is assumed to be zero.
		!
		!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
	
		k = 1
		
		select case(lnorm)
		
			case(1,4)
	
				if (lmax_comp /= 0) then
					k = k + 1
					f1(k) = sqr(3)
					f2(k) = 0.0d0
					k = k + 1
				endif
				
				do l=2, lmax_comp, 1
					k = k + 1
					f1(k) = sqr(2*l-1) * sqr(2*l+1) / dble(l)
					f2(k) = dble(l-1) * sqr(2*l+1) / sqr(2*l-3) / dble(l)
					do m=1, l-2, 1
						k = k+1
						f1(k) = sqr(2*l+1) * sqr(2*l-1) / sqr(l+m) / sqr(l-m)
                				f2(k) = sqr(2*l+1) * sqr(l-m-1) * sqr(l+m-1) &
                  				 	/ sqr(2*l-3) / sqr(l+m) / sqr(l-m) 
					enddo
					k = k+1
					f1(k) = sqr(2*l+1) * sqr(2*l-1) / sqr(l+m) / sqr(l-m)
                			f2(k) = 0.0d0
					k = k + 1
				enddo
			
			case(2)
			
				if (lmax_comp /= 0) then
					k = k + 1
					f1(k) = 1.0d0
					f2(k) = 0.0d0
					k = k + 1
				endif
				
				do l=2, lmax_comp, 1
					k = k + 1
					f1(k) = dble(2*l-1) /dble(l)
					f2(k) = dble(l-1) /dble(l)
					do m=1, l-2, 1
						k = k+1
						f1(k) = dble(2*l-1) / sqr(l+m) / sqr(l-m)
                  				f2(k) = sqr(l-m-1) * sqr(l+m-1) / sqr(l+m) / sqr(l-m)
					enddo
					k = k+1
					f1(k) = dble(2*l-1) / sqr(l+m) / sqr(l-m)
                  			f2(k) = 0.0d0
					k = k + 1
				enddo
			
			case(3)
		
				do l=1, lmax_comp, 1
					k = k + 1
					f1(k) = dble(2*l-1) /dble(l)
					f2(k) = dble(l-1) /dble(l)
					do m=1, l-1, 1
						k = k+1
						f1(k) = dble(2*l-1) / dble(l-m)
                  				f2(k) = dble(l+m-1) / dble(l-m)
					enddo
					k = k + 1
				enddo

		end select
	
		lmax_old = lmax_comp
		norm_old = lnorm
	
	endif
	
	!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
	!
	!	Do special case of lmax_comp = 0
	!
	!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
	
	if (lmax_comp == 0) then
	
		select case(lnorm)
			case(1,2,3);	pm2 = 1.0d0
			case(4);	pm2 = 1.0d0 / sqrt(4.0d0*pi)
		end select
		
		if (present(sampling)) then
		
			if (sampling == 1) then
				griddh(1:n, 1:n) = cilm(1,1,1) * pm2
			else
				griddh(1:n, 1:2*n) = cilm(1,1,1) * pm2
			endif
			
		else
		
			griddh(1:n, 1:n) = cilm(1,1,1) * pm2
		
		endif
	
		return
	
	endif
	
	!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
	!
	!	Determine Clms one l at a time by intergrating over latitude.
	!	
	!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
	
	call dfftw_plan_dft_c2r_1d(plan, nlong, coef(1:nlong/2+1), grid(1:nlong), FFTW_MEASURE)

	i_eq = n/2 + 1	! Index correspondong to zero latitude

	do i=1, i_eq - 1, 1
	
		i_s = 2*i_eq -i
	
		theta = pi * dble(i-1)/dble(n)
		z = cos(theta)
		u = sqrt( (1.0d0-z) * (1.0d0+z) )

		coef(1:lmax+1) = dcmplx(0.0d0,0.0d0)
		coef0 = 0.0d0
		coefs(1:lmax+1) = dcmplx(0.0d0,0.0d0)
		coef0s = 0.0d0
		
		select case(lnorm)
			case(1,2,3);	pm2 = 1.0d0
			case(4);	pm2 = 1.0d0 / sqrt(4.0d0*pi)
		end select

		tempr =  cilm(1,1,1) * pm2
		coef0 = coef0 + tempr
		coef0s = coef0s + tempr 	! symsign is always 1 for l=m=0
				
		k = 2
		pm1 =  f1(k) * z * pm2
		tempr = cilm(1,2,1) * pm1
		coef0 = coef0 + tempr
		coef0s = coef0s - tempr 	! symsign = -1
				
		do l=2, lmax_comp, 1
			l1 = l+1
			k = k+l
			p = f1(k) * z * pm1 - f2(k) * pm2
			tempr = cilm(1,l1,1) * p
			coef0 = coef0 + tempr
			coef0s = coef0s + tempr * symsign(k)
			pm2 = pm1
			pm1 = p
		enddo
				
		select case(lnorm)
			case(1,2);	pmm = sqr(2) * scalef
			case(3);	pmm = scalef
			case(4);	pmm = sqr(2) * scalef / sqrt(4.0d0*pi)
		end select
				
		rescalem = 1.0d0/scalef
		kstart = 1
			
		do m = 1, lmax_comp-1, 1
			
			m1 = m+1
			rescalem = rescalem * u
			kstart = kstart+m+1
					
			select case(lnorm)
				case(1,4)
					pmm = phase * pmm * sqr(2*m+1) / sqr(2*m)
					pm2 = pmm
				case(2)
					pmm = phase * pmm * sqr(2*m+1) / sqr(2*m)
					pm2 = pmm / sqr(2*m+1)
				case(3)
					pmm = phase * pmm * dble(2*m-1)
					pm2 = pmm
			end select
			
			tempc = dcmplx(cilm(1,m1,m1), - cilm(2,m1,m1)) * pm2
			coef(m1) = coef(m1) + tempc
			coefs(m1) = coefs(m1) + tempc
			! symsign(kstart) = 1
										
			k = kstart+m+1
	   		pm1 = z * f1(k) * pm2
	   				
	   		tempc = dcmplx(cilm(1,m1+1,m1), - cilm(2,m1+1,m1)) * pm1
	   		coef(m1) = coef(m1) + tempc	
			coefs(m1) = coefs(m1) - tempc
			! symsign = -1
	   				
			do l = m+2, lmax_comp, 1
				l1 = l+1
				k = k + l
				p = z * f1(k) * pm1 - f2(k) * pm2
				pm2 = pm1
       				pm1 = p
       				tempc = dcmplx(cilm(1,l1,m1), - cilm(2,l1,m1)) * p
				coef(m1) = coef(m1) + tempc
				coefs(m1) = coefs(m1) + tempc * symsign(k)
			enddo
					
			coef(m1) = coef(m1) * rescalem
			coefs(m1) = coefs(m1) * rescalem
					
		enddo			
								
		rescalem = rescalem * u
				
         	select case(lnorm)
            		case(1,4);	pmm = phase * pmm * sqr(2*lmax_comp+1) / sqr(2*lmax_comp) * rescalem
            		case(2);	pmm = phase * pmm / sqr(2*lmax_comp) * rescalem
            		case(3);	pmm = phase * pmm * dble(2*lmax_comp-1) * rescalem
        	end select
         			
        	tempc = dcmplx(cilm(1,lmax_comp+1,lmax_comp+1), - cilm(2,lmax_comp+1,lmax_comp+1)) * pmm
        	coef(lmax_comp+1) = coef(lmax_comp+1) + tempc
		coefs(lmax_comp+1) = coefs(lmax_comp+1) + tempc
		! symsign = 1
		
		coef(1) = dcmplx(coef0,0.0d0)
		coef(2:lmax+1) = coef(2:lmax+1)/2.0d0
		
		if (present(sampling)) then
			if (sampling == 2) then
				coef(lmax+2:2*lmax+3) = dcmplx(0.0d0,0.0d0)	
			endif
		endif
               			
               	call dfftw_execute(plan)	! take fourier transform
               	griddh(i,1:nlong) = grid(1:nlong)
               	
               	if (i /= 1) then	! don't compute value for south pole.
               		coef(1) = dcmplx(coef0s,0.0d0)
			coef(2:lmax+1) = coefs(2:lmax+1)/2.0d0
		
			if (present(sampling)) then
				if (sampling == 2) then
					coef(lmax+2:2*lmax+3) = dcmplx(0.0d0,0.0d0)	
				endif
			endif
              	
               		call dfftw_execute(plan)	! take fourier transform
               		griddh(i_s,1:nlong) = grid(1:nlong)
               	endif
			
	enddo
	
	! Finally, do equator
	
	z = 0.0d0
	u = 1.0d0

	coef(1:lmax+1) = dcmplx(0.0d0,0.0d0)
	coef0 = 0.0d0
	
	select case(lnorm)
		case(1,2,3);	pm2 = 1.0d0
		case(4);	pm2 = 1.0d0 / sqrt(4.0d0*pi)
	end select
	
	coef0 = coef0 + cilm(1,1,1) * pm2
				
	k = 2
				
	do l=2, lmax_comp, 2
		l1 = l+1
		k = k+l
		p = - f2(k) * pm2
		pm2 = p
		coef0 = coef0 + cilm(1,l1,1) * p
		k = k + l + 1
	enddo
				
	select case(lnorm)
		case(1,2);	pmm = sqr(2) * scalef
		case(3);	pmm = scalef
		case(4);	pmm = sqr(2) * scalef / sqrt(4.0d0*pi)
	end select
				
	rescalem = 1.0d0/scalef
	kstart = 1
			
	do m = 1, lmax_comp-1, 1
				
		m1 = m + 1
		kstart = kstart+m+1
					
		select case(lnorm)
			case(1,4)
				pmm = phase * pmm * sqr(2*m+1) / sqr(2*m)
				pm2 = pmm
			case(2)
				pmm = phase * pmm * sqr(2*m+1) / sqr(2*m)
				pm2 = pmm / sqr(2*m+1)
			case(3)
				pmm = phase * pmm * dble(2*m-1)
				pm2 = pmm
		end select
					
		coef(m1) = coef(m1) + dcmplx(cilm(1,m1,m1), &
				- cilm(2,m1,m1)) * pm2
					
		k = kstart+m+1 
	   				
		do l = m+2, lmax_comp, 2
			l1 = l+1
			k = k + l
			p = - f2(k) * pm2
			coef(m1) = coef(m1) + dcmplx(cilm(1,l1,m1), &
				- cilm(2,l1,m1)) * p
			pm2 = p
                 	k = k + l + 1
		enddo
					
	enddo			
				
        select case(lnorm)
            	case(1,4);	pmm = phase * pmm * sqr(2*lmax_comp+1) / sqr(2*lmax_comp)
            	case(2);	pmm = phase * pmm / sqr(2*lmax_comp)
            	case(3);	pmm = phase * pmm * dble(2*lmax_comp-1)
        end select
         			
        coef(lmax_comp+1) = coef(lmax_comp+1) + dcmplx(cilm(1,lmax_comp+1,lmax_comp+1), &
				- cilm(2,lmax_comp+1,lmax_comp+1)) * pmm
		
	coef(1) = dcmplx(coef0,0.0d0)
	coef(2:lmax+1) = coef(2:lmax+1) * rescalem /2.0d0
	
	if (present(sampling)) then
		if (sampling == 2) then
			coef(lmax+2:2*lmax+3) = dcmplx(0.0d0,0.0d0)	
		endif
	endif
	
        call dfftw_execute(plan)	! take fourier transform
                
	griddh(i_eq,1:nlong) = grid(1:nlong)

	call dfftw_destroy_plan(plan)
				
end subroutine MakeGridDH

